### PR TITLE
Add Metadata filtering on Pgvector

### PR DESCRIPTION
**Pull request by humcqc**
- Implement Metadata filtering on Pgvector with 3 implementations to store Metadata: JSON, JSONB, COLUMNS
- Add indexes on metadata 
- use Langchain4j metadata filter test
- Change tests to @QuarkusTest. QuarkusUnitTest extension does not support ParametrizedTest with custom classes as argument. to check with quarkus team.

Not sure if this should be in quarkus-langchain4j or in langchain4j ?
PR open for discussion.

### CODE REVIEW

**Comment by geoand** [comment]
Thanks!

I'll let @jmartisk review this

**Comment by humcqc** [comment]
yes, will cleanup after review and comments. I know i should not sync with the github site, but too late

**Comment by jmartisk** [comment]
I'll try to look into it asap, I'm not sure I'll have enough time before leaving until Tuesday.
Maybe @sebastienblanc would like to have a look too, as he implemented the Pgvector store.

**Comment by jmartisk** [comment]
I really do like it, very nice job! But yeah a lot of this looks like it could be moved to upstream langchain4j, which currently doesn't support metadata filtering on pgvector.
@langchain4j what do you think?
It could also benefit from the strategies (json, jsonb, columns) that you introduce for storing metadata

**Comment by langchain4j** [comment]
Yeah, would be awesome to have `Filter` mapping logic in https://github.com/langchain4j/langchain4j ! And the rest probably as well :)

**Comment by humcqc** [comment]
ok will move it to langchain4j, and adapt this one.

**Comment by sebastienblanc** [comment]
Will be happy to review this PR and also the parts moving to @langchain4j

**Comment by humcqc** [comment]
last commit integrates : https://github.com/langchain4j/langchain4j/pull/851

**Comment by jmartisk** [comment]
I can't get this to work somehow...
Does this add some new requirements on the application to specify some properties?
On startup I get:
```
Caused by: org.postgresql.util.PSQLException: Unknown type vector.
	at org.postgresql.jdbc.PgPreparedStatement.setPGobject(PgPreparedStatement.java:537)
	at org.postgresql.jdbc.PgPreparedStatement.setObject(PgPreparedStatement.java:1049)
	at io.agroal.pool.wrapper.PreparedStatementWrapper.setObject(PreparedStatementWrapper.java:291)
	at dev.langchain4j.store.embedding.pgvector.PgVectorEmbeddingStore.addAllInternal(PgVectorEmbeddingStore.java:324)
	... 26 more
```

So I tried adding 
```
quarkus.datasource.jdbc.additional-jdbc-properties.datatype.vector=com.pgvector.PGvector
```

in the application's properties, but then I get 

```
2024-04-03 14:26:27,559 WARN  [io.agr.pool] (agroal-11) Datasource '<default>': Unable to load the class com.pgvector.PGvector responsible for the datatype vector
(...)
java.lang.RuntimeException: Sql statement issue: org.postgresql.util.PSQLException: Unable to load the class com.pgvector.PGvector responsible for the datatype vector 
Last query: init
	at dev.langchain4j.store.embedding.pgvector.PgVectorEmbeddingStore.initTable(PgVectorEmbeddingStore.java:172)
	at dev.langchain4j.store.embedding.pgvector.PgVectorEmbeddingStore.<init>(PgVectorEmbeddingStore.java:81)
	at io.quarkiverse.langchain4j.pgvector.PgVectorEmbeddingStore.<init>(PgVectorEmbeddingStore.java:43)
	at io.quarkiverse.langchain4j.pgvector.runtime.PgVectorEmbeddingStoreRecorder.lambda$embeddingStoreFunction$1(PgVectorEmbeddingStoreRecorder.java:26)
	at io.quarkiverse.langchain4j.pgvector.PgVectorEmbeddingStore_NcsXENBovnGlSJZ1aaFjoohrtKM_Synthetic_Bean.createSynthetic(Unknown Source)
	... 36 more
```
Then I tried adding

```
quarkus.class-loading.parent-first-artifacts=com.pgvector:pgvector
```

but I am back at the first error

**Comment by humcqc** [comment]
interesting, does the rag-with-pgvector integration test work for you ?

**Comment by sebastienblanc** [comment]
@jmartisk Are you using a pg container that has pgvector ? It should mention it in the log if not https://github.com/quarkiverse/quarkus-langchain4j/blob/main/pgvector/runtime/src/main/java/io/quarkiverse/langchain4j/pgvector/PgVectorEmbeddingStore.java#L114

**Comment by humcqc** [comment]
@sebastienblanc should be automatic now with https://github.com/quarkiverse/quarkus-langchain4j/pull/399

**Comment by jmartisk** [comment]
Right, the integration test is failing too.
The image isn't the problem, it's using `pgvector/pgvector:pg16`

**Comment by jmartisk** [comment]
When running the integration tests I get (this is the relevant part)

```
2024-04-03 15:12:42,956 INFO  [tc.pgvector/pgvector:pg16] (build-41) Container pgvector/pgvector:pg16 is starting: f1c5cdc29c85b6a1e002dad4a4a1f1b457439d927cc514868338ce6920bfb0bf
2024-04-03 15:12:46,368 INFO  [tc.pgvector/pgvector:pg16] (build-41) Container pgvector/pgvector:pg16 started in PT3.496847295S
2024-04-03 15:12:46,369 INFO  [tc.pgvector/pgvector:pg16] (build-41) Container is started (JDBC URL: jdbc:postgresql://localhost:46373/quarkus?loggerLevel=OFF)
2024-04-03 15:12:46,369 INFO  [io.qua.dev.pos.dep.PostgresqlDevServicesProcessor] (build-41) Dev Services for PostgreSQL started.
2024-04-03 15:12:46,370 INFO  [io.qua.dat.dep.dev.DevServicesDatasourceProcessor] (build-41) Dev Services for default datasource (postgresql) started - container ID is f1c5cdc29c85
2024-04-03 15:12:51,787 INFO  [ai.djl.uti.Platform] (main) Found matching platform from: jar:file:///home/jmartisk/.m2/repository/ai/djl/huggingface/tokenizers/0.26.0/tokenizers-0.26.0.jar!/native/lib/tokenizers.properties
2024-04-03 15:12:52,054 WARN  [io.agr.pool] (main) Datasource '<default>': JDBC resources leaked: 0 ResultSet(s) and 1 Statement(s)
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 14.37 s <<< FAILURE! -- in org.acme.example.openai.RAGWithMetadataFilterTest
[ERROR] org.acme.example.openai.RAGWithMetadataFilterTest -- Time elapsed: 14.37 s <<< ERROR!
java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:638)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptBeforeAllMethod(QuarkusTestExtension.java:706)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:285)
	at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:252)
	at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:605)
	at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:655)
	... 1 more
Caused by: java.lang.RuntimeException: org.postgresql.util.PSQLException: Unknown type vector.
	at dev.langchain4j.store.embedding.pgvector.PgVectorEmbeddingStore.addAllInternal(PgVectorEmbeddingStore.java:345)
	at dev.langchain4j.store.embedding.pgvector.PgVectorEmbeddingStore.addInternal(PgVectorEmbeddingStore.java:294)
	at dev.langchain4j.store.embedding.pgvector.PgVectorEmbeddingStore.add(PgVectorEmbeddingStore.java:210)
	at dev.langchain4j.store.embedding.pgvector.PgVectorEmbeddingStore.add(PgVectorEmbeddingStore.java:36)
	at io.quarkiverse.langchain4j.pgvector.PgVectorEmbeddingStore_NcsXENBovnGlSJZ1aaFjoohrtKM_Synthetic_ClientProxy.add(Unknown Source)
	at org.acme.example.EmbeddingStoreIngestor.embed(EmbeddingStoreIngestor.java:29)
	at org.acme.example.EmbeddingStoreIngestor.ingest(EmbeddingStoreIngestor.java:22)
	at org.acme.example.EmbeddingStoreIngestor_Observer_ingest_UItmqUFF_K3-3VnNeybAPBBp55c.notify(Unknown Source)
	at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:346)
	at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:328)
	at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:82)
	at io.quarkus.arc.runtime.ArcRecorder.fireLifecycleEvent(ArcRecorder.java:155)
	at io.quarkus.arc.runtime.ArcRecorder.handleLifecycleEvents(ArcRecorder.java:106)
	at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy(Unknown Source)
	... 8 more
Caused by: org.postgresql.util.PSQLException: Unknown type vector.
	at org.postgresql.jdbc.PgPreparedStatement.setPGobject(PgPreparedStatement.java:537)
	at org.postgresql.jdbc.PgPreparedStatement.setObject(PgPreparedStatement.java:1049)
	at io.agroal.pool.wrapper.PreparedStatementWrapper.setObject(PreparedStatementWrapper.java:291)
	at dev.langchain4j.store.embedding.pgvector.PgVectorEmbeddingStore.addAllInternal(PgVectorEmbeddingStore.java:324)
	... 22 more

```

**Comment by humcqc** [comment]
ok , with fresh pg container i have the same issue.
But when i add:
```
quarkus.class-loading.parent-first-artifacts=com.pgvector:pgvector
```
It fixes the issue.

**Comment by jmartisk** [comment]
I had to add postgresql too,
```
quarkus.class-loading.parent-first-artifacts=com.pgvector:pgvector,org.postgresql:postgresql
```
to fix it in dev mode and prod mode (the `samples/chatbot` application) but it doesn't help with the integration tests.
I'm done for today, I'll give it some more thought later...

**Comment by humcqc** [comment]
Thanks @jmartisk this one point me to some interesting cascaded dependency issue. I will investigate and come back to you. not sure we can use parent first class loader in our case.

**Comment by humcqc** [comment]
I've change the way we handle connection settings. some weird behavior at the first usage of the store in certain condition.
Should be ok. 
I still need to find a better way when the db settings and the table is created by external script.

**Comment by humcqc** [comment]
@jmartisk, @sebastienblanc  I think with last commit it's ok, you don't need anymore to add the properties, i used a listener to add pgvector settings at connection creation. And it should work in all the case for quarkus. I've added a it with flyway and static embeddings table creation. Tell me if it's ok for you

**Comment by humcqc** [comment]
@jmartisk , @sebastienblanc, should be ok. If you can give a try to validate your external project. I've added 2 integrations tests for rag integration. one with automatic table creation and another with flyway. For the connection settings I use a pool listener to configure pgvector settings. Then getConnection  just give datasource connection. I've tested with internal test and external project. And with fresh pg containers. First time I have this behavior.

File: pgvector/runtime/src/main/java/io/quarkiverse/langchain4j/pgvector/runtime/PgVectorEmbeddingStoreConfig.java [hunk 1]
```diff
@@ -48,4 +52,56 @@ public interface PgVectorEmbeddingStoreConfig {
...
+        /**
+         * Metadata type:
+         * <ul>
+         * <li>COLUMNS: for static metadata, when you know in advance the list of metadata
+         * <li>JSON: For dynamic metadata, when you don't know the list of metadata that will be used.
+         * <li>JSONB: Same as JSON, but stored in a binary way. Optimized for query on large dataset.
+         * </ul>
+         * <p>
+         * Default value: JSON
+         */
```
Comment (jmartisk) [inline pgvector/runtime/src/main/java/io/quarkiverse/langchain4j/pgvector/runtime/PgVectorEmbeddingStoreConfig.java:73]:
```suggestion
        /**
         * Metadata type:
         * <ul>
         * <li>COLUMNS: for static metadata, when you know in advance the list of metadata fields. In this case,
         * you should also override the `quarkus.langchain4j.pgvector.metadata.definition` property to define the right columns.
         * <li>JSON: For dynamic metadata, when you don't know the list of metadata fields that will be used.
         * <li>JSONB: Same as JSON, but stored in a binary way. Optimized for query on large dataset. In this case,
         * you should also override the `quarkus.langchain4j.pgvector.metadata.definition` property to change the
         * type of the `metadata` column to JSONB.
         * </ul>
         * <p>
         * Default value: JSON
         */
```

File: pgvector/runtime/src/main/java/io/quarkiverse/langchain4j/pgvector/runtime/PgVectorEmbeddingStoreConfig.java [hunk 2]
```diff
@@ -48,4 +52,56 @@ public interface PgVectorEmbeddingStoreConfig {
...
+        @WithDefault("metadata JSON NULL")
+        List<String> definition();
+
+        /**
+         * Metadata Indexes, list of fields to use as index
+         * example:
+         * <ul>
+         * <li>JSON: metadata or (metadata->'key'), (metadata->'name'), (metadata->'age')
+         * <li>JSONB: (metadata_b->'key'), (metadata_b->'name'), (metadata_b->'age')
```
Comment (jmartisk) [inline pgvector/runtime/src/main/java/io/quarkiverse/langchain4j/pgvector/runtime/PgVectorEmbeddingStoreConfig.java:90]:
```suggestion
         * <li>JSONB: (metadata->'key'), (metadata->'name'), (metadata->'age')
```

**Comment by jmartisk** [comment]
Works for me!
I won't give a +1 yet because we need to wait for this to be in a released version of langchain4j, but given the current status of this PR plus the current status of https://github.com/langchain4j/langchain4j/pull/851, I think it's good.
Thanks so much, this is great.

**Comment by langchain4j** [comment]
@geoand I have to re-review it. How urgent is it?

**Comment by geoand** [comment]
Not terrible urgent from point of view, but it does seem like a very contribution

**Comment by humcqc** [comment]
For this one i will create a new PR when langchain will be released with : https://github.com/langchain4j/langchain4j/pull/851

**Comment by humcqc** [comment]
Waiting for langchain4J 0.31.0

**Comment by langchain4j** [comment]
@humcqc will be released next week

**Comment by humcqc** [comment]
Too many conflicts.
Replaced by https://github.com/quarkiverse/quarkus-langchain4j/pull/619
This PR will be closed soon

**Pull request closed without merging**